### PR TITLE
adding curve gauge connector

### PR DESCRIPTION
**Pull request by leckylao**
adding curve gauge connector;
remove trailing spaces;

### CODE REVIEW

File: contracts/connectors/curve_gauge.sol [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    uint _amt = getUint(getId, amt);
+    address curveGaugeAddr = CurveGaugeMapping(getCurveGaugeMappingAddr())
+      .gaugeMapping(bytes32(stringToBytes32(gaugePoolName)));
+    require(curveGaugeAddr != address(0), "wrong-gauge-pool-name");
+    IGauge gauge = IGauge(curveGaugeAddr);
+    TokenInterface crv_token = TokenInterface(address(gauge.crv_token()));
+    TokenInterface rewarded_token = TokenInterface(address(gauge.rewarded_token()));
+    Balances memory balances;
+
+    _amt = _amt == uint(-1) ? TokenInterface(address(gauge.lp_token())).balanceOf(address(this)) : _amt;
```
Comment (thrilok209) [inline contracts/connectors/curve_gauge.sol]: For `uint(-1)` case, we need to get the balance deposited in the staking/gauge contract. Right?

File: contracts/connectors/curve_gauge.sol [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    balances.rewardAmt = sub(balances.finalRewardBal, balances.intialRewardBal);
+
+    setUint(setId, _amt);
+    setUint(setIdCRVReward, balances.crvRewardAmt);
+    setUint(setIdReward, balances.rewardAmt);
+
+    emitLogWithdraw(gaugePoolName, _amt, getId, setId, setIdCRVReward, setIdReward);
+  }
+
+  function emitLogWithdraw(string memory gaugePoolName, uint _amt, uint getId, uint setId, uint setIdCRVReward, uint setIdReward) internal {
```
Comment (thrilok209) [inline contracts/connectors/curve_gauge.sol]: Add one more event like `LogClaimedReward()` which emits both crv and reward token amount. So, we create stats if we want to.

File: contracts/connectors/curve_gauge.sol [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    emit LogWithdraw(gaugePoolName, _amt, getId, setId, setIdCRVReward, setIdReward);
```
Comment (thrilok209) [inline contracts/connectors/curve_gauge.sol]: After adding `LogClaimedReward()` event, we can remove `setIdCRVReward` and `setIdReward` from `LogWithdraw` event.

File: contracts/connectors/curve_gauge.sol (hunk 2 above)
Comment (thrilok209) [reply contracts/connectors/curve_gauge.sol]: With `setIdCRVReward` and `setIdReward`

File: contracts/connectors/curve_gauge.sol [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+  /**
+  * @dev Withdraw LP Token.
+    * @param gaugePoolName gauge pool name.
+    * @param amt LP token amount.
+    * @param getId Get token amount at this ID from `InstaMemory` Contract.
+    * @param setId Set token amount at this ID in `InstaMemory` Contract.
+    * @param setIdCRVReward Set CRV token reward amount at this ID in `InstaMemory` Contract.
+  */
+  function withdraw(
```
Comment (thrilok209) [inline contracts/connectors/curve_gauge.sol]: Lets keep only one `withdraw()` function.

File: contracts/connectors/curve_gauge.sol [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    bytes memory _eventParamWithdraw = abi.encode(gaugePoolName, _amt, getId, setId, setIdCRVReward);
+    emitEvent(_eventCodeWithdraw, _eventParamWithdraw);
+  }
+
+  /**
+  * @dev Claim Reward.
+    * @param gaugePoolName gauge pool name.
+    * @param setId Set CRV reward amount at this ID in `InstaMemory` Contract.
+  */
+  function claimReward(
```
Comment (thrilok209) [inline contracts/connectors/curve_gauge.sol]: Same for `claimReward()` also. We can remove this function and keep the above function

File: contracts/connectors/curve_gauge.sol (hunk 4 above)
Comment (leckylao) [reply contracts/connectors/curve_gauge.sol]: As discussed one for with reward token, the other one for no reward token

File: contracts/connectors/curve_gauge.sol (hunk 5 above)
Comment (leckylao) [reply contracts/connectors/curve_gauge.sol]: As discussed one for with reward token, the other one for no reward token

File: contracts/connectors/curve_gauge.sol (hunk 4 above)
Comment (thrilok209) [reply contracts/connectors/curve_gauge.sol]: But still in no reward token function, user will be get the reward token and we are not tracking it.

File: contracts/connectors/curve_gauge.sol (hunk 1 above)
Comment (thrilok209) [reply contracts/connectors/curve_gauge.sol]: @leckylao is this correct?

Comment (thrilok209) [reply contracts/connectors/curve_gauge.sol]: Here, we are getting ERC20 balance, not deposited balance. Right?

Comment (thrilok209) [reply contracts/connectors/curve_gauge.sol]: Check in deposit function also. Line: 106. How can this be same in both the deposit and withdraw function?

File: contracts/connectors/curve_gauge.sol (hunk 4 above)
Comment (leckylao) [reply contracts/connectors/curve_gauge.sol]: image

Please check line 150. Now the LogClaimedReward event will track rewards

**Comment by leckylao** [comment]
image

File: contracts/connectors/curve_gauge.sol [hunk 6]
```diff
@@ -52,7 +52,7 @@ contract GaugeHelper is DSMath, Stores{
   }
 }
 
-contract CurveGauge is GaugeHelper {
+contract CurveGaugeEvent is GaugeHelper {
```
Comment (leckylao) [inline contracts/connectors/curve_gauge.sol:54]: why call event?

Comment (thrilok209) [reply contracts/connectors/curve_gauge.sol:54]: We can rename to `GaugeEventHelpers`

**Comment by leckylao** [comment]
image

**Review by thrilok209 — COMMENTED** [review]
Looks good to me. We can deploy it.

**Comment by leckylao** [comment]
Please also update and run the test after code changes to ensure nothing breaks

**Review by thrilok209 — APPROVED** [review]
Looks good

**Pull request merged**